text/0000-non-exhaustive.md

```diff
@@ -0,0 +1,482 @@
+value.
+
+# Detailed design
+
+An attribute `#[non_exhaustive]` is added to the language, which will (for now)
+fail to compile if it's used on anything other than an enum or struct
+definition.
+
+## Enums
+
+Within the crate that defines the enum, this attribute is essentially ignored,
+so that the current crate can continue to exhaustively match the enum. The
+justification for this is that any changes to the enum will likely result in
+more changes to the rest of the crate. Consider this example:
+
+```rust
+use std::error::Error as StdError;
+
+#[non_exhaustive]
+pub enum Error {
+    Message(String),
+    Other,
+}
+impl StdError for Error {
+    fn description(&self) -> &str {
+        match *self {
+            Message(ref s) => s,
+            Other => "other or unknown error",
+        }
+    }
+}
+```
+
+It seems undesirable for the crate author to use a wildcard arm here, to
+ensure that an appropriate description is given for every variant. In fact, if
+they use a wildcard arm in addition to the existing variants, it should be
+identified as dead code, because it will never be run.
+
+Outside the crate that defines the enum, users should be required to add a
+wildcard arm to ensure forward-compatibility, like so:
+
+```rust
+use mycrate::Error;
+
+match error {
+    Message(ref s) => ...,
+    Other => ...,
+    _ => ...,
+}
+```
+
+And it should *not* be marked as dead code, even if the compiler does mark it as
+dead and remove it.
+
+## Structs
+
+Like with enums, the attribute is essentially ignored in the crate that defines
+the struct, so that users can continue to construct values for the struct.
+However, this will prevent downstream users from constructing values or
+exhaustively matching values, because fields may be added to the struct in the
+future.
+
+For example, using our `Config` again:
+
+```rust
+#[non_exhaustive]
+pub struct Config {
+    pub window_width: u16,
+    pub window_height: u16,
+}
+```
+
+We can still construct our config within the defining crate like so:
+
+```rust
+let config = Config { window_width: 640, window_height: 480 };
+```
+
+And we can even exhaustively match on it, like so:
+
+```rust
+if let Ok(Config { window_width, window_height }) = load_config() {
+    // ...
+}
+```
+
+But users outside the crate won't be able to construct their own values, because
+otherwise, adding extra fields would be a breaking change.
+
+Users can still match on `Config`s non-exhaustively, as usual:
+
+```rust
+let &Config { window_width, window_height, .. } = config;
+```
+
+But without the `..`, this code will fail to compile. Additionally, the user
+won't be able to perform any functional-record-updates like the below:
+
+```rust
+let val = Config {
+    window_width: 640,
+    window_height: 480,
+    ..Default::default()
+};
+```
+
+Because there's no guarantee that the remaining fields will satisfy the
+requirements (in this case, `Default`).
```

Comment:
I don't think this is correct.  If a library provides a struct that's `Default`, then it's a breaking change for it to no longer implement `Default`, regardless of whether it's non-exhaustive.  FRU definitely needs to be prohibited for now, but because the new field could be private (not currently supported), not because the field might not have a default.

(Of course, that FRU limitation [could be fixed](https://github.com/rust-lang/rfcs/pull/1806#issuecomment-270881093) in a hypothetical future RFC.)